### PR TITLE
ARROW-13683: [R] Test Windows UCRT R

### PR DESCRIPTION
**Pull request by wjones127**

### CODE REVIEW

**Comment by pitrou** [comment]
I removed "WIP" from the title otherwise the GHA jobs won't run :-)

File: .github/workflows/r.yml [hunk 1]
```diff
@@ -241,14 +241,17 @@ jobs:
...
     runs-on: windows-latest
     if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
     timeout-minutes: 60
     strategy:
       fail-fast: false
       matrix:
-        rtools: [35, 40]
+        config:
+        - { rtools: 35, rversion: "3.6" }
+        - { rtools: 40, rversion: "4.1" }
```
Comment (nealrichardson) [inline .github/workflows/r.yml:253]:
```suggestion
        - { rtools: 40, rversion: "release" }
```

**Comment by wjones127** [comment]
> I removed "WIP" from the title otherwise the GHA jobs won't run :-)

Thanks @pitrou! I was planning on enabling that once I saw the relevant CI job succeeded in my fork:
https://github.com/wjones127/arrow/runs/4232096168?check_suite_focus=true

**Comment by wjones127** [comment]
It seems to be using the UCRT R binary: 

```
* using R Under development (unstable) (2021-11-09 r81170 ucrt)
```
https://github.com/apache/arrow/runs/4240709248?check_suite_focus=true#step:16:487

Thought it doesn't look like we have it set up with the UTF-8 charset as default:

```
* using platform: x86_64-w64-mingw32 (64-bit)
* using session charset: ISO8859-1
```

https://github.com/apache/arrow/runs/4240709248?check_suite_focus=true#step:16:488

I think this is sufficient for now, but once the UCRT starts to get in use we may need to make some additional changes to test UTF-8 support on Windows. (Right now I don't see much activity from any other R package to test this.)

**Comment by nealrichardson** [comment]
@jeroen what do you think, does that look correct to you?

**Review by jonkeane — APPROVED** [review]
This looks good to me, is there anything else we need to do before merging it?

**Comment by wjones127** [comment]
I think this should be good for now. I expect once R 4.2 is released we will revisit the UTF support with the new RTools.

**Comment by jeroen** [comment]
You need to switch to `r-lib/actions/setup-r@v2`. The `master` branch of `r-lib/actions/setup-r` is dead.
Also you can use `r: devel` instead of `r:devel-ucrt` as they are synonymous now.

**Comment by wjones127** [comment]
I switched to `devel` instead of `devel-ucrt`, but that seems to have caused an issue: [processx](https://cran.r-project.org/web/packages/processx/index.html) package not yet available for `r-devel`, but is available for `r-devel-UCRT`.

**Comment by jonkeane** [comment]
Is this ready for review (or merging?)

**Comment by wjones127** [comment]
> Is this ready for review (or merging?)

No, I think we need to wait on processx to be released for `r-devel` on Windows (or find some workaround) before this will pass in CI.

**Comment by jeroen** [comment]
processx is available for all versions, I think this was just a random network failure or something.

**Comment by pitrou** [comment]
@wjones127 What is the status on this?

**Comment by wjones127** [comment]
> @wjones127 What is the status on this?

TBH I was hoping that the upstream processx issue would resolve itself. I can return to this next week and see if there is a fix we can implement to work around it. (It sounds like Jeroen thinks there is.)

**Comment by wjones127** [comment]
@jonkeane processx still has no Window R-devel binary, so I've switched the CI to install from source for now. This is ready for review + merge now.

File: .github/workflows/r.yml [hunk 2]
```diff
@@ -233,14 +233,18 @@ jobs:
...
+        # TODO: Once R 4.2 comes out we can switch to devel + 4.2
+        - { rtools: 40, rversion: "devel-ucrt" }
```
Comment (nealrichardson) [inline .github/workflows/r.yml:247]: See https://github.com/r-lib/actions/blob/v2-branch/setup-r/src/installer.ts#L683-L686

Also we should test on the toolchain that CRAN uses. So we may continue to use rtools40 to build the C++ library but should use rtools42 for R >= 4.2.

```suggestion
        - { rtools: 42, rversion: "devel" }
```

Comment (wjones127) [reply .github/workflows/r.yml:247]: Thanks!

File: .github/workflows/r.yml [hunk 3]
```diff
@@ -252,31 +256,31 @@ jobs:
...
       - name: Download artifacts
-        if: ${{ matrix.rtools == 40 }}
+        if: ${{ matrix.config.rtools == 40 }}
         uses: actions/download-artifact@v2
         with:
           name: libarrow-rtools40-mingw64.zip
           path: r/windows
       - name: Download artifacts
-        if: ${{ matrix.rtools == 40 }}
+        if: ${{ matrix.config.rtools == 40 }}
```
Comment (nealrichardson) [inline .github/workflows/r.yml:283]:
```suggestion
        if: ${{ matrix.config.rtools == 42 }}
```

**Review by nealrichardson — APPROVED** [review]

**Pull request closed without merging**